Add BeginCreate and BeginUpdate REST hooks

thockin opened the pull request:
These hooks return a "cleanup" func which is called when the top-level
operation completes, with an indicator of which result.

This is to enable much simpler handling of allocations in Service's REST
implementation, in particular.

Some discussion in https://github.com/kubernetes/kubernetes/pull/95967

This also adds tests for the almost totally untested Decorator,
AfterCreate, and AfterUpdate hooks.  My gift to you as thanks for considering
these hooks :)


/kind feature

```release-note
NONE
```

thockin commented:
cc @aojea

aojea commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -500,6 +535,14 @@ func (e *Store) Update(ctx context.Context, name string, objInfo rest.UpdatedObj
+				}
+				cleanup = fn
+			}
+
```
don't we need to set `success = true` before returning in L563?

aojea replied:
ahh this runs inside `e.Storage.GuaranteedUpdate(`
sorry

aojea commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -146,14 +146,25 @@ type Store struct {
 	// specific cases where storage of the value is not appropriate, since
 	// they cannot be watched.
 	Decorator ObjectFunc
+
```
The decorator is no longer the "last" exit hook , should we make it more explicit?

thockin replied:
Good point, will add to comments

thockin replied:
re-pushed

thockin committed:
Add BeginCreate and BeginUpdate REST hooks

These hooks return a "cleanup" func which is called when the top-level
operation completes, with an indicator of which result.

This is to enable much simpler handling of allocations in Service's REST
implementation, in particular.

Some discussion in https://github.com/kubernetes/kubernetes/pull/95967

This also adds tests for the almost totally untested Decorator,
AfterCreate, and AfterUpdate hooks.

liggitt commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -602,9 +654,19 @@ func (e *Store) Update(ctx context.Context, name string, objInfo rest.UpdatedObj
 			return nil, false, err
 		}
 	}
+	success = true // used in the defer'ed cleanup
```
same question about success being false in a persisted-yet-AfterCreate-failed scenario

aojea replied:
we already have a bug because we do deallocate despite the object is persisted, because there a repair controller that will reallocate the resource from the object stored, but during that time another service can take that resource
https://github.com/kubernetes/kubernetes/pull/93199#issuecomment-665157865

liggitt commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -500,6 +535,14 @@ func (e *Store) Update(ctx context.Context, name string, objInfo rest.UpdatedObj
+				if err != nil {
+					return nil, nil, err
+				}
+				cleanup = fn
```
Overwriting `cleanup` here and calling it outside GuaranteedUpdate means that if GuaranteedUpdate is called multiple times, only the last invocation of BeginCreate will get its returned function invoked.

If we're allocating inside BeforeCreate, doesn't that mean:
* allocations of user-specified ports will fail on retries because they conflict with previously allocated things from previous invocations?
* auto-allocated ports from previous invocations will leak?

liggitt replied:
this could be mitigated by calling `cleanup` on retry before invoking it again and overwriting `cleanup`, e.g.:

```go
if e.BeginCreate != nil {
  if cleanup != nil {
    // we were invoked previously, indicate that attempt did not succeed
    cleanup(false)
    cleanup = nil
  }
  fn, err := e.BeginCreate(obj, newCreateOptionsFromUpdateOptions(options))
  ...
  cleanup = fn
}
```

liggitt replied:
basically, I would expect the invariant that every invocation of BeforeCreate/BeforeUpdate for a given request has its returned function called exactly once, and that call occurs prior to any reinvocations of BeforeCreate/BeforeUpdate for that request

thockin replied:
good point.  I did call it internally at first but then the confusion around AfterUpdate and Decorator hit me.  Before I make this more complex, let's sort that...

thockin replied:
Tests are green on additional commits to change other hooks to not return error.

thockin replied:
Now added a test for the inner-retry loop.

liggitt commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -544,6 +587,15 @@ func (e *Store) Update(ctx context.Context, name string, objInfo rest.UpdatedObj
+			if err != nil {
+				return nil, nil, err
+			}
+			cleanup = fn
```
same question about overwriting cleanup from previous invocations

liggitt commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -602,9 +654,19 @@ func (e *Store) Update(ctx context.Context, name string, objInfo rest.UpdatedObj
+	return &metav1.CreateOptions{
+		DryRun:       in.DryRun,
+		FieldManager: in.FieldManager,
+	}
```
I can't remember if populating the type meta is important, but patch does when converting PatchOptions to `{Create,Update}Options`:

https://github.com/kubernetes/kubernetes/blob/0ad06e991ab0d0708593e3e6415c521c6b76b7d1/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go#L647-L671

thockin replied:
ACK

thockin commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -146,14 +146,25 @@ type Store struct {
+	// commit/cleanup function which will be called at the end of operation
+	// (after AfterCreate and Decorator), indicating via the argument whether
+	// the operation succeeded.  Almost nobody should use this hook.
+	BeginCreate func(obj runtime.Object, options *metav1.CreateOptions) (func(bool), error)
```
ACK.  I guess we should also pass ctx to the cleanup func?  I don't think using a capture there is a good idea.

liggitt replied:
yeah, threading it to both seems right to me

thockin replied:
Fixed in next push

thockin commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -146,14 +146,25 @@ type Store struct {
+	// commit/cleanup function which will be called at the end of operation
+	// (after AfterUpdate and Decorator), indicating via the argument whether
+	// the operation succeeded.  Almost nobody should use this hook.
+	BeginUpdate func(obj, old runtime.Object, options *metav1.UpdateOptions) (func(bool), error)
```
ACK.  Same Q as above

thockin commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -391,6 +416,7 @@ func (e *Store) Create(ctx context.Context, obj runtime.Object, createValidation
 			return nil, err
 		}
 	}
+	success = true // used in the defer'ed cleanup
```
So this was not clear to me.  Why do we allow After* and Decorator to return an error at all?  If they fail, it seems like we hit a pretty awful path - operation failed but object exists?

Given how little those are used - can we just make them not return an error?

I can flip the sequence so that cleanup happens before these and just pretend this weirdness doesn't exist, of course....

liggitt replied:
> Why do we allow After* and Decorator to return an error at all? If they fail, it seems like we hit a pretty awful path - operation failed but object exists?

Agree. "Why do we even have this lever?"

> Given how little those are used - can we just make them not return an error?

That would be my inclination. @smarterclayton?

thockin replied:
ICYMI - I added commits that implement this - nobody (at least in-tree) uses those hooks to return errors.

thockin commented:
Pushed with some changes applied.  Need to figure out what After* and Decorator failure semantics we really want, and whether to pass ctx to cleanup.

fedebongio commented:
@thockin is your plan to get this one into 1.20?

fedebongio commented:
/triage accepted

thockin commented:
This seems exceedingly unlikely for 20

thockin commented:
/retest

thockin committed:
Make REST After* funcs not return error

thockin committed:
Make REST Decorator funcs not return error

thockin commented:
OK, I have pushed now with new commits to make After* and Decorator not return errors.  The last commit is still in progress but I wanted to push before I run out of time today.

liggitt commented on pkg/registry/core/service/storage/storage.go:
```diff
@@ -241,6 +235,4 @@ func (r *GenericREST) defaultOnReadService(service *api.Service) error {
 		}
 	}
-
-	return nil
```
I'm relieved to see that this didn't actually drop any error-returning cases

thockin replied:
Yes, that made this plausible.

deads2k replied:
> I'm relieved to see that this didn't actually drop any error-returning cases

Yep, I had that thought as I started too. :)

liggitt commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -381,15 +408,17 @@ func (e *Store) Create(ctx context.Context, obj runtime.Object, createValidation
+	// The operation has succeeded.  Call the cleanup function if there is one,
+	// and then make sure the defer doesn't call it again.
+	if cleanup != nil {
+		cleanup(true)
```
if `cleanup()` panics here, the current code will call it again in the defer (applies in Update paths as well)

thockin replied:
Copy it to a temp var and call that?

thockin replied:
fixed in next push

liggitt reviewed (commented):
I like the direction. I'm mostly left wondering what to do with storage errors like timeouts, which are an error, but don't positively mean the object wasn't persisted. Maybe we just have to caveat these hooks with "you better have a reconcile process in place"

thockin commented:
Re-pushed with all comments addressed, I think.

There are "WIP" commits to change the signature of After* and Decorator, on the presumption that this will fly.  It makes sense to me, but I don't know what out-of-tree impact that will have, really.

This adds tests for the Decorator and After* Hooks (which were more or less untested) and also tests the update-retry path.

TODO: Is `ObjectFunc` a useful type?  If so should it change (as proposed here) or should we define new types for the new signatures (name suggestions welcome)?  Happy to go either way.

TODO: The manual-cleanup-on-success could be simplified to just `defer` if the Create path used an "inner" function.  E.g.:
```
    func Create(...) {
        err := func() {
            success := false
            cleanup := BeginCreate()
            defer cleanup(success)()
            // do create
            success = true
            return nil
        }()
        AfterCreate()
        Decorator()
    }
```
Update is already stuctured this way, but I chose to do them both in the same style.  Open to debate.

This is *NOT* a priority for 1.20, but even a nod to "this seems fundamentally OK" means I can star implementing the Service REST usage of this.

thockin commented:
/retest

thockin commented:
So now that code-freeze is behind us, a high-level nod indicating "we aren't going to fundamentally shoot this down" would be awesome...

liggitt commented:
I'm +1 on `After{Create,Update,Delete}` and `Decorator` losing the ability to return an error

This has the potential to be much better for the service storage, and didn't really do violence to `generic/registry/store.go`

The main gap I see remaining is a timeout/network error coming back from the etcd create/update call that makes storage think the operation failed when it really succeeded (this isn't a new issue, it exists in service storage today). As long as anyone using this understands they need a reconciler to fixup cases like that, I think this is a significant improvement over what we have today.

liggitt commented:
daniel/clayton/david/wojtek should probably chime in now

thockin commented:
Thanks Jordan.

@lavalamp @deads2k @smarterclayton for a look. I have started doing the service cleanups in https://github.com/kubernetes/kubernetes/pull/96684

thockin commented:
Small wrinklee here, as I am working on using this.  Delete.

I thought I would use After Delete(), but it turns out that hook is called regardless of whether the operation is dryRun or not.  If I want service to deallocate IPs when the service is deleted, I don't want to do that on a dry-run.

I thought I might to the same begin/cleanup dance, but I need to delete IPs *after* finalizers, not in the Delete method (there's an open bug!).  With finalizers that can be (very) async to the actual Delete method call.  No good.

I know that AfterDelete is called at the right point in time, but it doesn't carry DeleteOptions or even `dryRun bool`.

So I am looking for input.  It seems reasonable (to me) to either add `dryRun bool` to AfterDelete() (or to all of the After* hooks?) - would that be acceptable?  Given that no other options have been needed so far (nobody uses the After* hooks in-tree) that seems simpler than adding CreatOptions, UpdateOptions, DeleteOptions.

Is there a better course?

thockin commented:
I added a new commit with the AfterDelete() change (and tests for it).  Open questions:
    * Should we pass the full DeleteOptions instead?
    * Should we do this for AfterCreate and AfterUpdate, too?

wojtek-t commented:
> daniel/clayton/david/wojtek should probably chime in now

I didn't look super carefully into it, but I'm fully supportive for the direction here.
Daniel/David/Clayton should look into it as they are more familiar with all corner cases.

thockin commented:
Pinging on this.  I am getting anxious about building too much more on this until I know it's viable.

@liggitt note the changes in delete-path

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -50,10 +50,10 @@ import (
+// ObjectFunc is a function to act on a given object. Such hooks may not fail,
+// and do not have any way to indicate an error.  An ObjectFunc may transform
+// the provided object.
+type ObjectFunc func(obj runtime.Object)
```
This change breaks OpenShift (we use errors from ObjectFuncs today).  Today these hooks can fail, why does this need to change now? Perhaps you need a different kind of hook?  Our decorators can return errors for instance: extreme edges, but they do so today.

thockin replied:
If they fail on a mutating operation, the object is persisted in etcd anyway, right?  That seems like a pretty awkward state to leave users.  What failures are really meaningful at this late stage of processing?

deads2k replied:
> If they fail on a mutating operation, the object is persisted in etcd anyway, right? That seems like a pretty awkward state to leave users. What failures are really meaningful at this late stage of processing?

Decorators can be used to place dynamic information on the return value and may fail.  While the data is persisted, returning the unmutated data back could return incomplete or incorrect information.  We've cleaned out most egregious examples, but at least one still exists.

thockin replied:
As discussed today, we agreed to keep this change.

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/decorated_watcher.go:
```diff
@@ -21,17 +21,18 @@ import (
 type decoratedWatcher struct {
 	w         watch.Interface
-	decorator ObjectFunc
+	decorator func(runtime.Object)
```
reading down, this was actually still an ObjectFunc.  Can we leave it as one.

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -146,14 +147,25 @@ type Store struct {
+	// commit/revert function which will be called at the end of operation,
+	// but before AfterCreate and Decorator, indicating via the argument
+	// whether the operation succeeded.  Almost nobody should use this hook.
+	BeginCreate func(ctx context.Context, obj runtime.Object, options *metav1.CreateOptions) (func(context.Context, bool), error)
```
nit: I have a issue with being able to read functions returning functions.  Can we name this function type please?  `type MutatingCallbackFunc func(context.Context, succeeded bool)`

deads2k replied:
nit: worth noting that the hook is not called if BeginCreate returns an error, so it cannot do all compensation logic.

thockin replied:
Are you attached to that name?  I went with `FinishFunc` (Begin <-> Finish), but I'll go with your pref.

Will be in next push

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -337,6 +352,22 @@ func (e *Store) ListPredicate(ctx context.Context, p storage.SelectionPredicate,
+		defer func() {
+			// If we succeed, we will have already set cleanup to nil.
+			if cleanup != nil {
+				cleanup(ctx, false)
```
What do we suggest cleanup functions do if they fail to cleanup?

thockin replied:
Log/Event.  Have an async GC system (like service repair loops).  Unless we want to *really* plumb REAL transactions in here, there's got to be a turtle at the bottom that doesn't return an error.

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -162,7 +174,10 @@ type Store struct {
+	//FIXME: pass whe whole options instead?
+	//FIXME: Do the same for other After* hooks?
+	//FIXME: if so, get rid of ObjectFunc type?
+	AfterDelete func(obj runtime.Object, dryRun bool)
```
it seems a little weird that only this After func needs to be aware of dryRun.  Is this becasue there isn't a BeginDelete?

thockin replied:
Yes - because Delete is a split operation.  The Delete() method can complete, but the object isn't gone until either the grace period expires or some update removes a finalizer.

I'd be fine to plumb dryRun to other After* - it seems correct, but I wanted feedback.

thockin replied:
small bump here - you wanted the whole options passed to After*, which I agreed with.  The trick is that the split-delete means I don't have a DeleteOptions in all cases.  I can synthesize one from an UpdateOptions - fair?

thockin replied:
In the next push

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -500,6 +534,22 @@ func (e *Store) Update(ctx context.Context, name string, objInfo rest.UpdatedObj
 		doUnconditionalUpdate := newResourceVersion == 0 && e.UpdateStrategy.AllowUnconditionalUpdate()
 
 		if existingResourceVersion == 0 {
+			var cleanup func(context.Context, bool)
```
nit: can I get a `createCleanup` here and an `updateCleanup` down below for future me?

thockin replied:
ACK.  In next push :)

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -381,15 +412,18 @@ func (e *Store) Create(ctx context.Context, obj runtime.Object, createValidation
 	}
+	// The operation has succeeded.  Call the cleanup function if there is one,
+	// and then make sure the defer doesn't call it again.
+	if cleanup != nil {
```
why do we cleanup before we run AfterCreate and Decorator?

thockin replied:
Update's inner-loop.  I think we want to bound the begin/cleanup to that helper function.  Otherwise we have to call begin on the inside, publish the cleanup method to the outside function and call it there.

We *could* do that, it just seemed simpler to keep it contained.

thockin replied:
As discussed today, keeping this ordering

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -586,25 +668,30 @@ func (e *Store) Update(ctx context.Context, name string, objInfo rest.UpdatedObj
 
+// This is a helper to convert UpdateOptions to CreateOptions for the
+// create-on-update path.
+func newCreateOptionsFromUpdateOptions(in *metav1.UpdateOptions) *metav1.CreateOptions {
```
Can we get a test that ensures we're handling all fields? Something like we did for the `rest.CopyConfig`, that makes us look at this method every time we add a value to UpdateOptions or to CreateOptions.

thockin replied:
context https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/rest/config_test.go#L395-L493

thockin replied:
ACK.. It's a little gross, but it works.

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go:
```diff
@@ -531,6 +641,343 @@ func TestNoOpUpdates(t *testing.T) {
+		begin: func(_ context.Context, obj, _ runtime.Object, _ *metav1.UpdateOptions) (func(context.Context, bool), error) {
+			mile("BeginUpdate")
+			return func(context.Context, bool) {
+				mile("Cleanup")
```
I'd like these tests to expand to indicate whether they were called with success true or false.  I like the test.

thockin replied:
ACK.  Also will cover failure paths better

thockin reviewed (commented):
Ignoring code-level comments in favor of major issues.

Update is gnarly but not impossible.  If you so demand, I'll try to figure it out.

Errors from After and Decorators scare the pants off me.

@liggitt too :)

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -337,6 +352,22 @@ func (e *Store) ListPredicate(ctx context.Context, p storage.SelectionPredicate,
 
 // Create inserts a new item according to the unique key from the object.
 func (e *Store) Create(ctx context.Context, obj runtime.Object, createValidation rest.ValidateObjectFunc, options *metav1.CreateOptions) (runtime.Object, error) {
```
```go
func cleanNothing(ctx, bool){}

Create() {
  var cleanupFn cleanupFunc = cleanNothing
  success := false
  defer func(){
    cleanup(ctx, success)
  }

  if e.BeginCreate != nil{
    fn, err := e.BeginCreate
    if err != nil{
    }
    cleanFn = fn
  }

  // where the existing duplicate cleanup block is
  success = true
```

thockin replied:
Given the ordering we talked about, I can do something similar but not quite this.  I'm going to add it as a distinct commit.  If you like it, we keep it.  If not, we abandon just that.

thockin commented:
Did a quick chat with @deads2k.  Notes:

- OK to change Decorator and After*
- Add comments around how to use these properly
- OK with hook sequencing as-is
- pass *Options to After*, OK with new function signatures

I will iterate on this and the other smaller feedback in the coming weeks - not a rush.

thockin committed:
Remove ObjectFunc

thockin committed:
Change AfterDelete to take DeleteOptions

All of the After* hooks are called, whether the operation is dry-run or
not.  Create and Upate both have BeginCreate() hooks which know about
dry-run but delete doesn't, and can't (because of graceful deletion and
finalizers, the Delete() method may return and the object is actually
deleted later).

This adds an argument to AfterDelete(), and the others will follow.

This commit also adds tests for AfterDelete being called in the right
places.

thockin committed:
Change AfterCreate/Update to take Options

Also adds typedefs for these function signatures for code clarity.

thockin committed:
Use typedefs for Begin* functions

David asked for this for readability, even if not all other hooks do it.

thockin commented:
New push is up.  As far as I can tell, all comments are addressed.  I'm happy to make further changes if you like.

The `defer` logic has been reworked as its own commit which should be fixup'ed or dropped depending on how @deads2k feels about it.  https://github.com/kubernetes/kubernetes/pull/96393#discussion_r543807093

I'm not PERSONALLY keen on the function typedefs that get used just one time, but this is not my long-term maintenance area, so I defer readily.

PTAL

aojea commented:
/retest
I think that someone reported issues with the 100 job today 🤔 

k8s-infra-e2e-boskos-scale-23
W1217 19:58:57.560] Zone: us-east1-b
W1217 19:59:01.010] ERROR: (gcloud.compute.instance-groups.managed.list-instances) Some requests did not succeed:
W1217 19:59:01.010]  - Internal error. Please try again or contact Google Support. (Code: '5B6AE6C4AD5D7.69001D5.8F0B1933')

The Basel job failed because of the test
TestWatchFileChanged

aojea commented:
/retest

this is a known one https://github.com/kubernetes/kubernetes/issues/96803
>  Kubernetes e2e suite: [sig-api-machinery] API priority and fairness should ensure that requests can be classified by testing flow-schemas/priority-levels expand_more

aojea commented:
Hmm failed twice the same test?


Kubernetes e2e suite: [sig-api-machinery] API priority and fairness should ensure that requests can be classified by testing flow-schemas/priority-levels expand_less | 3s
-- | --
test/e2e/apimachinery/flowcontrol.go:48 Dec 23 10:42:23.371: matching user doesnt received UID for the testing priority-level and flow-schema vendor/github.com/onsi/ginkgo/internal/leafnodes/runner.go:113

/retest

aojea commented:
we are good now :+1:

deads2k commented on pkg/registry/core/service/storage/storage.go:
```diff
@@ -131,42 +131,36 @@ func (r *StatusREST) Update(ctx context.Context, name string, objInfo rest.Updat
 	service, ok := obj.(*api.Service)
 	if ok {
-		return r.defaultOnReadService(service)
+		r.defaultOnReadService(service)
```
this still works, but it's a little weird not to return. You do still want to return in this branch, right? 

maybe a type switch would be more normal

thockin replied:
ACK, type switch would be.

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go:
```diff
@@ -376,6 +368,256 @@ func TestStoreCreate(t *testing.T) {
+		// Handle any expected differences before we compare.
+
+		// Compare the results.
+		inBytes, err = json.Marshal(inBytes)
```
I feel a little silly for asking, but what does does marshalling bytes produce?

thockin replied:
good catch - I got it right in TestNewDeleteOptionsFromUpdateOptions

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go:
```diff
@@ -376,6 +368,256 @@ func TestStoreCreate(t *testing.T) {
+		}
+
+		// Handle any expected differences before we compare.
+		delete(inMap, "fieldManager")
```
give a comment here about delete options not having a fieldmanager field?

thockin replied:
ACK

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go:
```diff
@@ -376,6 +368,256 @@ func TestStoreCreate(t *testing.T) {
+		// Handle any expected differences before we compare.
+		delete(inMap, "fieldManager")
+		delete(outMap, "gracePeriodSeconds")
+		delete(outMap, "Preconditions")
```
this case looks odd. Did we lose a json tag in our structs somewhere?

thockin replied:
Good catch - it was `omitEmpty'ed`

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go:
```diff
@@ -376,6 +368,256 @@ func TestStoreCreate(t *testing.T) {
+				mile(fmt.Sprintf("FinishCreate(%v)", success))
+			}, nil
+		},
+		expectMilestones: []string{"BeginCreate", "FinishCreate(true)", "AfterCreate", "Decorator"},
```
thanks for the bool output

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go:
```diff
@@ -531,6 +773,390 @@ func TestNoOpUpdates(t *testing.T) {
+		milestones = append(milestones, s)
+	}
+
+	testCases := []struct {
```
I thought one of the semantics you wanted was that every attempt at an update got a separate finish call.  I don't think I see a test case checking for that semantic.  Given the likelihood of a refactor on the update code, I think it could help ensure the behavior sticks.  Is there a reasonable way to have the `NewTestGenericStoreRegistry` fail the guaranteed update?

thockin replied:
Look at TestStoreUpdateHooksInnerRetry() - it uses the TTLFunc hook to force a failure and retry on the inner loop, so you see cases like:

```
expectMilestones: []string{"BeginUpdate", "TTLError", "FinishUpdate(false)", "BeginUpdate", "TTL", "FinishUpdate(true)", "AfterUpdate", "Decorator"},
```

and 

```
expectMilestones: []string{"BeginUpdate", "TTLError", "FinishUpdate(false)", "BeginUpdate", "TTLError", "FinishUpdate(false)"},
```

I assume that handles what you are asking for?  Will resolve on that presumption.

deads2k replied:
> I assume that handles what you are asking for? Will resolve on that presumption.

so it is.  thank you.

deads2k commented:
lgtm other than fairly minor comments.  https://github.com/kubernetes/kubernetes/pull/96393#discussion_r554008797 is the only one that keeps me from tagging,  I'm not sure the test case is doing what it intends to do.

In the comments, future you/us may appreciate knowing that the ordering of `BeforeX` with respect to mutating admission versus canonicalization in a strategy.  I know the endpoints strategy makes use of it at least.

thockin commented:
I believe all comments are addressed.

deads2k commented:
> I believe all comments are addressed.

I agree, thanks.

/lgtm

aojea commented:
/retest

thockin commented:
/retest
